src/libAtomVM/nifs.c

```diff
@@ -3452,6 +3459,79 @@ static term nif_erlang_make_fun_3(Context *ctx, int argc, term argv[])
     return term_make_function_reference(module_term, function_term, arity_term, &ctx->heap);
 }
 
+static term nif_erlang_fun_info_2(Context *ctx, int argc, term argv[])
+{
+    UNUSED(argc);
+    term fun = argv[0];
+    VALIDATE_VALUE(fun, term_is_fun);
+    term key = argv[1];
+    VALIDATE_VALUE(key, term_is_atom);
+
+    term value;
+    switch (key) {
+        case MODULE_ATOM: {
+            const term *boxed_value = term_to_const_term_ptr(fun);
+            if (term_is_external_fun(fun)) {
+                term module_atom = boxed_value[1];
+
+                value = module_atom;
+            } else {
+                Module *module = (Module *) boxed_value[1];
+
+                value = module_get_name(module);
+            }
+            break;
+        }
+        case NAME_ATOM: {
+            if (term_is_external_fun(fun)) {
+                const term *boxed_value = term_to_const_term_ptr(fun);
+                term name_atom = boxed_value[2];
+
+                value = name_atom;
+            } else {
+                char *function_name = NULL;
+                size_t function_name_len = term_local_function_name(fun, ctx->global, &function_name);
+
+                value = make_list_from_utf8_buf((uint8_t *) function_name, function_name_len, ctx);
+            }
+            break;
+        }
+
+        case ARITY_ATOM:
+            if (term_is_external_fun(fun)) {
+                const term *boxed_value = term_to_const_term_ptr(fun);
+                term arity = boxed_value[3];
+
+                value = make_maybe_boxed_int64(ctx, term_to_int32(arity));
+            } else {
+                const term *boxed_value = term_to_const_term_ptr(fun);
+                Module *module = (Module *) boxed_value[1];
+                uint32_t fun_index = term_to_int32(boxed_value[2]);
+
+                uint32_t label, arity, n_freeze;
+                module_get_fun(module, fun_index, &label, &arity, &n_freeze);
+
+                value = make_maybe_boxed_int64(ctx, arity);
+            }
+            break;
+
+        case TYPE_ATOM:
+            value = term_is_external_fun(fun) ? EXTERNAL_ATOM : LOCAL_ATOM;
+            break;
+
+        default:
+            AVM_ABORT();
+    }
+
+    if (UNLIKELY(memory_ensure_free_opt(ctx, TUPLE_SIZE(2), MEMORY_CAN_SHRINK) != MEMORY_GC_OK)) {
+        RAISE_ERROR(OUT_OF_MEMORY_ATOM);
+    }
```

Comment:
I feel like there need to be better system for using those. GC invalidates any pointers to data on process' heap and some functions use it internally. Not only it makes several GC runs but also prevents allocating terms one by one.

Initially, I implemented both /1 and /2 as NIF but experienced problems of invalid pointers when I tried to turn a cstring into a list – it calls a GC so creating second list invalidates pointer to the first one.
I assume the correct way is to put the first one as a root but then the cstring -> list function needs to take two another parameters.
It's a hard problem but I thought about passing some struct that accumulates how much memory is needed and does it as one pass, something like adding an out parameter `GCTracker* tracker` that is used before final allocation. This necessarily splits memory calculation and memory allocation.